core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java

```diff
@@ -125,6 +135,47 @@ public void clusterChanged(ClusterChangedEvent event) {
                     }
                 }
             }
+
+            // check if we need to schedule a previous unscheduled (or mis-scheduled, due to a change in the setting)
+            // reroute because of delayed allocation.
+            StringBuilder sb = new StringBuilder();
+            boolean scheduled = false;
+            for (ShardRouting shard : event.state().routingTable().shardsWithState(ShardRoutingState.UNASSIGNED)) {
+                if (shard.primary()) {
+                    continue;
+                }
+                if (shard.unassignedInfo() == null) {
+                    continue;
+                }
+                final ShardId shardId = shard.shardId();
+                IndexMetaData indexMetaData = event.state().metaData().index(shard.getIndex());
+                long delayTimeoutSetting = shard.unassignedInfo().getAllocationDelayTimeoutSetting(settings, indexMetaData.getSettings());
+                if (delayTimeoutSetting > 0) {
+                    Tuple<Long, ScheduledFuture> registeredDelay = delayedShardsToReroute.get(shardId);
+                    if (registeredDelay == null || registeredDelay.v1() != delayTimeoutSetting) {
+                        if (registeredDelay != null) {
+                            FutureUtils.cancel(registeredDelay.v2());
+                        }
+                        TimeValue calculatedDelay = TimeValue.timeValueMillis(shard.unassignedInfo().getDelayAllocationExpirationIn(settings, indexMetaData.getSettings()));
+                        if (calculatedDelay.millis() > 0) {
+                            sb.append(shard.shardId()).append("/").append(shard.unassignedInfo().getReason()).append("/").append(calculatedDelay).append(",");
+                            scheduled = true;
+                            ScheduledFuture future = threadPool.schedule(calculatedDelay, ThreadPool.Names.SAME, new Runnable() {
+                                @Override
+                                public void run() {
+                                    delayedShardsToReroute.remove(shardId);
+                                    routingTableDirty = true;
+                                    reroute();
+                                }
+                            });
+                            delayedShardsToReroute.put(shard.shardId(), Tuple.tuple(delayTimeoutSetting, future));
+                        }
+                    }
+                }
+            }
+            if (scheduled) {
+                logger.info("delaying unassigned shard allocation, shards: {}", sb);
```

Comment:
this can very verbose (600 shards on a node). I'm doubting whether we should have this as debug and have an info level log saying "allocation of [X] unassigned shards delayed".